modules/afsocket/afinet-dest-failover.c

```diff
@@ -0,0 +1,443 @@
+                  evt_tag_str("next_failover_server", _get_hostname(self->current_server)),
+                  log_expr_node_location_tag(self->owner_expression));
+      return;
+    }
+
+  msg_warning("Current failover server is inaccessible, sending the messages to the next failover server",
+              evt_tag_str("next_failover_server", _get_hostname(self->current_server)),
+              log_expr_node_location_tag(self->owner_expression));
+}
+
+
+const gchar *
+afinet_dd_failover_get_hostname(AFInetDestDriverFailover *self)
+{
+  if (self->current_server == NULL)
+    {
+      return _get_hostname(_primary(self));
+    }
+
+  return _get_hostname(self->current_server);
+}
+
+void
+afinet_dd_failover_next(AFInetDestDriverFailover *self)
+{
+  if (self->failover_transport_mapper.transport_mapper == NULL) // not initialized
+    return;
+
+  if (!self->current_server)
+    {
+      self->current_server = _primary(self);
+      return;
+    }
+
+  _step_current_server_iterator(self);
+}
+
+void
+afinet_dd_failover_add_servers(AFInetDestDriverFailover *self, GList *failovers)
+{
+  self->servers = g_list_concat(self->servers, failovers);
+}
+
+void
+afinet_dd_failover_enable_failback(AFInetDestDriverFailover *self, gpointer cookie,
+                                   afinet_dd_failover_on_primary_available callback_function)
+{
+  self->on_primary_available_cookie = cookie;
+  self->on_primary_available_func = callback_function;
+}
+
+static void
+_afinet_dd_init_failback_handlers(AFInetDestDriverFailover *self)
+{
+  IV_TIMER_INIT(&self->timer);
+  self->timer.cookie = self;
+  self->timer.handler = _afinet_dd_failback_timer_elapsed;
+
+  IV_FD_INIT(&self->fd);
+  self->fd.cookie = self;
+  self->fd.handler_out = (void (*)(void *)) _afinet_dd_handle_tcp_probe_socket;
+}
+
+static void
+_afinet_dd_stop_failback_handlers(AFInetDestDriverFailover *self)
+{
+  if (iv_timer_registered(&self->timer))
+    iv_timer_unregister(&self->timer);
+
+  if (iv_fd_registered(&self->fd))
+    {
+      iv_fd_unregister(&self->fd);
+      close(self->fd.fd);
+    }
+}
+
+void
+afinet_dd_failover_init(AFInetDestDriverFailover *self, LogExprNode *owner_expr, FailoverTransportMapper *failover_transport_mapper)
+{
+  self->owner_expression = owner_expr;
+  self->failover_transport_mapper = *failover_transport_mapper;
+  self->current_server = NULL;
+  _afinet_dd_init_failback_handlers(self);
+}
+
+AFInetDestDriverFailover *
+afinet_dd_failover_new(const gchar *primary)
+{
+  AFInetDestDriverFailover *self = g_new0(AFInetDestDriverFailover, 1);
+  self->probe_interval = TCP_PROBE_INTERVAL_DEFAULT;
+  self->probes_required = SUCCESSFUL_PROBES_REQUIRED_DEFAULT;
+  self->servers = g_list_append(NULL, g_strdup(primary));
+  self->current_server = g_list_first(self->servers);
+  return self;
+}
+
+void
+afinet_dd_failover_free(AFInetDestDriverFailover *self)
+{
+  if (!self)
+    return;
+
+  _afinet_dd_stop_failback_handlers(self);
```

Comment:
This must be called in `deinit()` instead of `free()`, because you initialize handlers in `init()`.

If you create a destination but you forget to add it to a logpath, this will crash.

We don't call `init()` in that case, but `free()` is always called and ivykis crashes if you try to unregister a zero-initialized struct (ivykis prefers -1 values instead of 0, that's why we call `IV_FD_INIT` and `IV_TIMER_INIT`).